### ISSUE TITLE
GSTheme - Themes should have metadata indicating whether it is a "Dark" or "Light" theme...

### ISSUE THREAD

**Issue by gcasa** [gnustep/libs-gui#323]
Themes should make this distinction so that editors (and printing) can adjust the background and text color accordingly.   This should not simply be left up to the palette defined in each theme as this could vary.

**Comment by rmottola** [comment]
While trying to make GNUMail mode-savvy I am investigating behaviour on Mac and certain things are not linear.

When changing from light to dark mode, notification to dark mode is issued.
However, when an application is started and dark mode is already set, no notification is sent, so the application doesn't know what mode is running. On UIKit it is accessible apparently, but on AppKit I found out only a convoluted way, NSAppearance.

E.g. get the view Appearance through:
[[[self window] effectiveAppearance] name]

and then check the name of the appearance. Appearances are not really identical to our GSThemes, since they also contain context-dependent drawing styles (Vibrancy, etc)

[NSAppearanceNameAqua](apple-reference-documentation://hcAYhHqkga) vs [NSAppearanceNameDarkAqua](apple-reference-documentation://hc16jFKoV0)

but also e.g.
[NSAppearanceNameAccessibilityHighContrastAqua](apple-reference-documentation://hc-9w606bl) vs [NSAppearanceNameAccessibilityHighContrastDarkAqua](apple-reference-documentation://hcwc8BB47z)

and the same for Vibrancy.

A lot of code around just checks for AquaDark but that is clearly wrong, a decent hack might be to check Dark.

This is a little bit complicated to make compatible.
If we will support modes in themes we could append Dark to the name... and get something like GNUstep vs GNUstepDark or Sleek vs SleekDark

But for a theme only designed to be dark it needs Dark in the name. By coincidence it would work for ThinkDark but for another XXX theme maybe XXXDark should be returned as NSAppearanceName.

After we get of course NSAppearance and playing it along with themes.

**Comment by rmottola** [comment]
@fredkiefer indeed, we support in GUI e.g. labelColor and currently it is hard-wired to a specific color, for example, black.
I was thinking themes and thematic.

currently, Thematic does not support labelColor. I might just add it as other colors so one can set it.
This would all work for themes which decide to be light or dark.

but thinking forward to a theme supporting both light and dark, what should we do?
have two colors prepared and return the right one. E.g. labelColorLight, labelColorDark

I hope we can discuss that soon.

**Comment by rfm** [comment]
Currently GSTheme and Thematic.app support named colors with special states 'Selected' or 'Highlighted' to indicate how the named color varies depending on the state of the control.   The actual name of the color in the color list is determined by concatenating the color name and the state name, and the default value (without state) is used if there is no color defined for the current state of the widget.
It doesn't seem hard to extend that naming scheme for light and dark modes, as you suggest, by appending the current mode (if any),
At runtime, the theme would need to try various names in sequence until it gets a match (nameStateMode,  nameState, name).

**Comment by DTW-Thalion** [comment]
Coming to this from the code side, the two directions in the thread need different things, and I do not think it can be built until one is chosen.

The original request treats a theme as being dark or light, which is metadata about the theme. That part is cheap: `GSTheme` already reads `GSThemeVersion`, `GSThemeLicense` and `GSThemeAuthors` from the bundle's Info.plist, and a key alongside them plus an accessor is a few lines.

The naming scheme @rfm described, which @rmottola agreed with, implies the opposite: one theme serves both modes and the mode is runtime state, so a theme level dark or light flag would be the wrong thing to consult. That direction also has a head start, since `-colorNamed:state:` already resolves a name against a state, and a mode suffix extends a mechanism that exists rather than adding one.

The two can be combined, with metadata declaring which appearances a theme supports, light, dark or both, and the colour lookup resolving per mode. But that is a decision about the model rather than something to infer from the thread.

On the runtime side the foundations are already there: `NSAppearance` defines all four dark names, and `-effectiveAppearance` exists on `NSView` and `NSApplication`, so a theme can be told the current mode without new plumbing.

Which model do you want? I am happy to implement either, and the metadata piece can land first under either one if that helps.

**Comment by rmottola** [comment]
@DTW-Thalion indeed, there are two paths I want to make co-exist. Currently I put this "on hold" because of the lots of current work provided by you as well as the hard TypeSetter issue.

First, we already have dark themes, but they work poorly because nobody "knows" they are dark themes. I want to make them report as similar as dark appearance. Also Apple sends a notification when changing appearance Dark / Light and we should do that on theme change if switching from a light to dark one.

Second, a single theme could in long-term support multiple modes, similar to Apple's single system theme. We do not support that yet in thematic and I guess it needs further work, but I want to make sure that the first stage is not developed without keeping second one in mind.

For both I would set theme properties that Thematic should set and SystemPreferences should read.

We currently have no accessibility Appearances.
